jwt-post.md

```diff
@@ -0,0 +1,171 @@
+
+If you put your JWTs in a header, you don't need to worry about CSRF.
+You do need to worry about XSS, however.
+If someone can abuse XSS to steal your JWT, this person is able to impersonate you.
+
+In my 2016 presentation, I stated that "defense against CSRF is straightforward and durable."
+This statement was based on advice offered by the [Open Web Application Security Project (OWASP)](https://www.owasp.org/) at that time.
+Years later, defense against CSRF is still durable, but a little less straightforward.
+We'll come back to that in a minute.
+
+XSS, on the other hand, is something you need to constantly keep in mind.
+Each template you add could open up possibilities for XSS.
+The same holds for all those NPM packages you add to your front-end project, either directly or indirectly.
+
+My conclusion from this is that JWTs belong in a secure, HTTP-only cookie, and should be used in combination with preventive measures against CSRF.
+
+## Seeing is believing
+
+Because the proof of the pudding is in the eating, I wrote a simple front-end app and two back-end apps that demonstrate a session-based and JWT-based approach to authentication: https://github.com/ljpengelen/java-meetup-jwt
+
+With a simple `docker-compose` command, you can start three instances of either of the two back ends, a database, and an instance of nginx that serves the front end and acts as load balancer.
+You can open the front end in your browser, create an account, log in, and then stop some of the back-end instances with `docker stop`.
+
+In the case of the JWT-based back end, it doesn't matter which two instances you stop.
+In the case of the session-based back end, stopping the instance your connected to will terminate your session.
+
+# Measures against CSRF
+
+The OWASP has a [cheat sheet about measures against CSRF](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html).
+The applications mentioned above use two of those measures.
+First, they combat CSRF by checking the `Origin` and `Referer` headers.
+Second, they require that each request contains a secure random token, generated by a cryptographically secure random-number generator.
+Keeping track of the CSRF tokens in the front end is not completely straightforward, but it's an acceptable price to pay if you ask me.
+
+## Lifespan of a JWT
+
+Think about the following for a second: What happens to already issued JWTs when you change your credentials?
+What happens to already issued JWTs when you delete your account?
+In both scenarios, existing JWTs will remain valid.
+Without additional measures, JWTs remain valid until they expire or until the secret on the server is changed.
+If someone gets their hands on a token, it can be abused until it expires.
+If you want to invalidate a single token by changing the secret on the server, you invalidate all tokens.
+
+When should a JWT expire?
```

Comment:
Indeed this I think is the biggest downside/complexity of JWTs. There's no 'standard' way to do this, but there are some workarounds that people use like for example including the timestamp of when a password was last changed in the payload. Kind of a hassle, but it works.